### PR TITLE
Add primary, surrogate key, and grain glossary pages

### LINKED ISSUE #1396: Surrogate Keys
for Developer Blog
### Contact Details

@dave in slack

### This issue is being created to promote an approved Discussion.

- [X] I am promoting this issue from an approved Discussion and will link it below.

### Link to the Discussion this Issue is promoting.

#1303 

### If you're promoting this from an 'I want to write about...' Discussion, you can copy over the answers to the following questions. Otherwise, please answer these questions below

In a conversation with @mikegfuller and @matt-winkler, we were chatting about the pain of migrating from a SQL Server-style deployment to dbt with a specific eye towards maintaining monotonically increasing integer surrogate keys. While they were useful in the past (better join performance, coordination between loading Kimball-style dims and facts etc), they make your project stateful -- in order to add new data to model A, you need to know about the current state of the data in model A. This goes against the core principal of idempotency, and can lead you into very scary design patterns if you try to deploy this type of setup in dbt.

This article would advocate for derived surrogate keys, the power of the surrogate key dbt util, and to help you start to treat your models like cattle instead of pets (@randypitcherii 's words, not mine)

What is the main problem you are solving? What is your solution?
teaching folks coming from legacy tools how to rethink their assumptions about surrogate keys. The solution would be to derive the keys from the raw data in your warehouse in an idempotent way (dbt_utils.surrogate_key)

Why should the reader care about this problem? Why is your solution the right one? This should help form your specific target audience.
Trying to build surrogate keys the Old Way in dbt is a real pain, and does not actually help anyone. The target audience here are folks trying to do a 1:1 migration of Old Data to New Data -- now is the time to rethink your assumptions!

### Link to the initial outline we will develop through this Issue.

1. I don't think so -- defintions of key concepts will be covered in this article
2. not that i'm aware of -- we'll have some macros and the like, but nothing beyond that that a reader would need
3. yeah! most of the article will be about tradeoffs
4. the original surrogate key SQL magic post
5. natural keys all the way down

[link to the current blustfall](https://www.figma.com/file/uzd6EOCTrfZb3ATvcfMMo2/Surrogate-Key-Outline?node-id=0%3A1)

### LINKED ISSUE #1424: Glossary: Primary Key
### Contact Details

@KiraFuruichi

### This issue is being made to create a marketing-approved piece.

- [X] I work for dbt Labs and am creating this issue for a marketing-approved piece.

### Link to the initial outlines we will develop through this Issue.

https://docs.google.com/document/d/1R8_YbZQ0AsWT7ICxboT4pSxtS23lCYpEhEsqLOhnRZc/edit#

Took a mix of John's content brief + David's primary key devblog post (https://docs.getdbt.com/blog/primary-key-testing#does-your-warehouse-support-primary-keys) to essentially go a level-deeper in the latter. There's pretty heavy overlap when it comes to the DWH support for this stuff w/ David's post, so would love people's thoughts on this!

### LINKED ISSUE #1459: Update Snowflake section of "What's a primary key and why do we test them?" devblog post
### Contributions

- [X] I have read the contribution docs, and understand what's expected of me.

### Link to the page on docs.getdbt.com requiring updates

https://docs.getdbt.com/blog/primary-key-testing#snowflake-primary-keys

### What part(s) of the page would you like to see updated?

> "Note that Snowflake primary keys enforce a not null constraint, but not uniqueness - so you’ll still want to test whether your primary key column values are actually unique (see testing section at the bottom)."

After doing some research, [Snowflake supports the declaration of primary keys](https://docs.snowflake.com/en/sql-reference/constraints-overview.html), but will only enforce non-nullness for the `NOT NULL` constraint, so we should change this sentence. 

### Additional information

Will update this in the PR for adding the Primary Key glossary page

cc @krevitt @jasnonaz @johnblust

### PR DESCRIPTION
**Pull request by johnblust**
Adds primary key, surrogate key, and data grain glossary pages. This PR does not include internal linking. Will create a new branch for that so it's easier to review.

### CODE REVIEW

File: website/docs/terms/primary-key.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+---
+id: primary-key
+title: Primary Key
+displayText: Primary key  
+hoverSnippet: A primary key is a non-null column in a database object that uniquely identifies each row.
+---
+
+A primary key is a non-null column in a database object that uniquely identifies each row. Primary keys take the form of a natural or </Term id="surrogate-key" />. It’s important to note that for each <Term id="table" /> or <Term id="view" /> in your database, there must only be one primary key column per database object.
```
Comment (KiraFuruichi) [inline website/docs/terms/primary-key.md]: The surrogate key isn't showing up as a link/hover--should it be like that?

File: website/docs/terms/surrogate-key.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
+---
+id: surrogate-key
+title: Surrogate Key
+displayText: surrogate key  
+hoverSnippet: A surrogate key is a unique identifier derived from the data itself. It often takes the form of a hashed value of multiple columns that will create a uniqueness constraint for each row.
+---
+
+A surrogate key is a unique identifier derived from the data itself. It often takes the form of a hashed value of multiple columns that will create a uniqueness constraint for each row. You will need to create a surrogate key for every table that doesn't have a natural <Term id="primary-key" />.
```
Comment (KiraFuruichi) [inline website/docs/terms/surrogate-key.md:8]: Is there now way to lowercase the link to primary key glossary page?

File: website/docs/terms/surrogate-key.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+:::
+
+## How surrogate keys are created
+
+In analytics engineering, you can generate surrogate keys using a hashing method of your choice. Remember, in order to truly create a uniqueness constraint on a database object, you’ll need to hash the fields together that _make each row unique_; when you generate a correct surrogate key for a dataset, you’re really establishing the true <Term id="grain" /> of that dataset.
+
+Let’s take this to an example. Below, there is a table you pull from an ad platform that collects `calendar_date`, `ad_id`, and some performance columns.
+
+<table>
```
Comment (KiraFuruichi) [inline website/docs/terms/surrogate-key.md:31]: Table is missing the first row's date!

File: website/docs/terms/surrogate-key.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+## Conclusion
+
+Surrogate keys are unique row identifiers that are created by using columns in a database object to create a uniqueness constraint on the data. To create a surrogate key, you will use a cryptographic algorithm usually in the form of the MD5 function to hash together fields that create a uniqueness constraint on the dataset. Ultimately, surrogate keys are a great way to create unique row identifiers for database objects that lack them naturally and allow folks to easily identify the grain of the data.
+
+## Further reading
+
+Want to learn more about keys, dbt, and everything in-between? Check out the following:
+
+* [Glossary: Primary Keys](https://docs.getdbt.com/terms/primary-key)
+* [Generating Surrogate Keys across Warehouses](https://docs.getdbt.com/blog/sql-surrogate-keys)
```
Comment (KiraFuruichi) [inline website/docs/terms/surrogate-key.md]:
```suggestion
* [Generating surrogate keys across warehouses](https://docs.getdbt.com/blog/sql-surrogate-keys)
```

creating consistent casing

File: website/docs/terms/surrogate-key.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+## Conclusion
+
+Surrogate keys are unique row identifiers that are created by using columns in a database object to create a uniqueness constraint on the data. To create a surrogate key, you will use a cryptographic algorithm usually in the form of the MD5 function to hash together fields that create a uniqueness constraint on the dataset. Ultimately, surrogate keys are a great way to create unique row identifiers for database objects that lack them naturally and allow folks to easily identify the grain of the data.
+
+## Further reading
+
+Want to learn more about keys, dbt, and everything in-between? Check out the following:
+
+* [Glossary: Primary Keys](https://docs.getdbt.com/terms/primary-key)
```
Comment (KiraFuruichi) [inline website/docs/terms/surrogate-key.md]:
```suggestion
* [Glossary: Primary keys](https://docs.getdbt.com/terms/primary-key)
```
consistent casing

**Review by KiraFuruichi — CHANGES_REQUESTED** [review]
Looks great, thanks for getting this up!! Only a few minor things to change.

I was wondering if we could slip in a fix to [this section of this blog post](https://docs.getdbt.com/blog/primary-key-testing#snowflake-primary-keys). Link to the issue for it can be found [here](https://github.com/dbt-labs/docs.getdbt.com/issues/1459).

The second paragraph of the Snowflake section should read, "Note that Snowflake primary keys are purely declarative--neither uniqueness nor non-nullness constraints are enforced. However, Snowflake supports a separate `not null` constraint that can be applied to column that is enforced. Regardless, you'll still want to ensure your primary key column values are actually unique (see testing section at the bottom.)"

**Comment by KiraFuruichi** [comment]
Follow-up question: should the glossary page for grain be called "Data Grain" or just "Grain"?

File: website/docs/terms/surrogate-key.md (hunk 2 above)
Comment (johnblust) [reply website/docs/terms/surrogate-key.md:8]: Yeah I noticed that, meant to fix it in this PR, but I fixed it in the branch I believe

File: website/docs/terms/primary-key.md (hunk 1 above)
Comment (johnblust) [reply website/docs/terms/primary-key.md]: Weird not sure why it was doing that. I just typed the same thing out and now it works

File: website/docs/terms/surrogate-key.md (hunk 3 above)
Comment (johnblust) [reply website/docs/terms/surrogate-key.md:31]: Good catch, looks like it got dropped when I converted everything to md

**Comment by johnblust** [comment]
Thanks for the review, Kira! I made all changes and added the Snowflake section fix. Data grain is the proper name according to Google so I chose to go with that as the page title name.

**Review by KiraFuruichi — APPROVED** [review]

**Pull request merged**